ports/zephyr/zephyr_display.c

```diff
@@ -0,0 +1,333 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/display.h>
+#include <stdlib.h>
+
+#include "modzephyr.h"
+#include "py/runtime.h"
+
+#if defined(CONFIG_DISPLAY)
+
+#if DT_HAS_COMPAT_STATUS_OKAY(zephyr_displays)
+#define ENUMERATE_DISPLAY_DEVS(node_id, prop, idx) DEVICE_DT_GET(DT_PROP_BY_IDX(node_id, prop, idx)),
+
+static const struct device *zephyr_display_devices[DT_ZEPHYR_DISPLAYS_COUNT] = {
+    DT_FOREACH_PROP_ELEM(DT_COMPAT_GET_ANY_STATUS_OKAY(zephyr_displays), displays, ENUMERATE_DISPLAY_DEVS)
+};
+#elif DT_HAS_CHOSEN(zephyr_display)
+static const struct device *zephyr_display_devices[] = {
+    DEVICE_DT_GET(DT_CHOSEN(zephyr_display)),
+};
+#else
+static const struct device *zephyr_display_devices[1];
+#endif
+
+const mp_obj_type_t zephyr_display_type;
+
+typedef struct _zephyr_display_obj_t {
+    mp_obj_base_t base;
+    uint8_t id;
+} zephyr_display_obj_t;
+
+static void zephyr_display_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    zephyr_display_obj_t *self = self_in;
+    mp_printf(print, "Display(%d)", self->id);
+}
+
+static mp_obj_t zephyr_display_make_new(const mp_obj_type_t *type,
+    size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    int id = mp_obj_get_int(args[0]);
+
+    if (id < 0 || id >= DT_ZEPHYR_DISPLAYS_COUNT) {
+        mp_raise_ValueError("Invalid display ID");
+        return mp_const_none;
+    }
+    if (!device_is_ready(zephyr_display_devices[id])) {
+        mp_raise_msg(&mp_type_RuntimeError, "Display is not ready");
+        return mp_const_none;
+    }
+
+    zephyr_display_obj_t *self = mp_obj_malloc(zephyr_display_obj_t, type);
+    self->id = id;
+    return MP_OBJ_FROM_PTR(self);
+}
+
+static mp_obj_t zephyr_display_write(size_t n_args, const mp_obj_t *args_in) {
+    zephyr_display_obj_t *self = MP_OBJ_TO_PTR(args_in[0]);
+    struct display_capabilities capabilities;
+    int ret;
+    size_t len;
+    mp_buffer_info_t bufinfo;
+    int x = mp_obj_get_int(args_in[2]);
+    int y = mp_obj_get_int(args_in[3]);
+    int sx;
+    int sy;
+
+    mp_get_buffer_raise(args_in[1], &bufinfo, MP_BUFFER_READ);
+
+    display_get_capabilities(zephyr_display_devices[self->id], &capabilities);
+    sx = capabilities.x_resolution;
+
+    if (n_args > 4) {
+        sx = mp_obj_get_int(args_in[4]);
+    }
+    sy = bufinfo.len / (sx * DISPLAY_BITS_PER_PIXEL(capabilities.current_pixel_format) / 8);
+
+    if (n_args > 5) {
+        sy = mp_obj_get_int(args_in[5]);
+    }
+
+    len = sx * sy * DISPLAY_BITS_PER_PIXEL(capabilities.current_pixel_format) / 8;
+
+    if (len > bufinfo.len) {
+        mp_raise_ValueError("Buffer is shorter than size requires");
```

Comment:
This and other errors should use `MP_ERROR_TEXT` so it doesn't explode the build if/when MicroPython's error text compression is enabled, eg:

```c
mp_raise_ValueError(MP_ERROR_TEXT("Buffer is shorter than size requires"));
```